docs/searching.md

```diff
@@ -249,4 +249,216 @@ var query = searcher.CreateQuery();
+
+Basic example
+```csharp
+var searcher = myIndex.Searcher;
+var results = searcher.CreateQuery()
+ .Field("Address", "Hills")
+ .And()
+ .WithFacet("Address") // Get facets of the Address field
+ .Execute();
+
+var addressFacetResults = results.GetFacet("Address"); // Returns the facets for the specific field Address
+
+/* 
+* Example value
+* Label: Hills, Value: 2
+* Label: Hollywood, Value: 10
+*/
+
+var hillsValue = addressFacetResults.WithFacet("Hills"); // Gets the IFacetValue for the facet Hills
+```
+
+Filtered value example
+```csharp
+var searcher = myIndex.Searcher;
+var results = searcher.CreateQuery()
+ .Field("Address", "Hills")
+ .And()
+ .Facet("Address", "Hills") // Get facets of the Address field
+ .Execute();
+
+var addressFacetResults = results.GetFacet("Address"); // Returns the facets for the specific field Address
+
+/* 
+* Example value
+* Label: Hills, Value: 2 <-- As Hills was the only filtered value we will only get this facet
+*/
+
+var hillsValue = addressFacetResults.WithFacet("Hills"); // Gets the IFacetValue for the facet Hills
+```
+
+MaxCount example
+```csharp
+var searcher = myIndex.Searcher;
+var results = searcher.CreateQuery()
+ .Field("Address", "Hills")
+ .And()
+ .WithFacet("Address") // Get facets of the Address field
+ .Execute();
+
+var addressFacetResults = results.GetFacet("Address"); // Returns the facets for the specific field Address
+
+/* 
+* Example value
+* Label: Hills, Value: 2
+* Label: Hollywood, Value: 10
+* Label: London, Value: 12
+*/
+
+results = searcher.CreateQuery()
+ .Field("Address", "Hills")
+ .And()
+ .WithFacet("Address") // Get facets of the Address field
+    .MaxCount(2) // Gets the top 2 results (The facets with the highest value)
+ .Execute();
+
+addressFacetResults = results.GetFacet("Address"); // Returns the facets for the specific field Address
+
+/* 
+* Example value (Notice only 2 values are present)
+* Label: Hollywood, Value: 10
+* Label: London, Value: 12
+*/
+```
+
+FacetField example
+```csharp
+// Setup
+
+// Create a config
+var facetsConfig = new FacetsConfig();
+
+// Set the index field name to facet_address. This will store facets of this field under facet_address instead of the default $facets. This requires you to use FacetField in your Facet query. (Only works on string facets).
+facetsConfig.SetIndexFieldName("Address", "facet_address");
+
+services.AddExamineLuceneIndex("MyIndex",
+    // Set the indexing of your fields to use the facet type
+    fieldDefinitions: new FieldDefinitionCollection(
+        new FieldDefinition("Address", FieldDefinitionTypes.FacetFullText)
+        ),
+    // Pass your config
+    facetsConfig: facetsConfig
+    );
+
+
+var searcher = myIndex.Searcher;
+var results = searcher.CreateQuery()
+ .Field("Address", "Hills")
+ .And()
+ .WithFacet("Address") // Get facets of the Address field
+    .FacetField("address_facet")
```

Comment:
Should this be `facet_address` as used in other part of code?

Reply:
To my understanding you can name it whatever you like as long as it's indexed with the same name as when you search. All facets are by default indexed under the $facets field.

Reply:
Okay, I tried this simple example https://lucenenet.apache.org/docs/4.8.0-beta00008/api/Lucene.Net.Demo/Lucene.Net.Demo.Facet.SimpleFacetsExample.html with the current Examine v3, but it seems I need to index/define the facets in the index.

I noticed this PR has some changes regarding the document, but do you know if we can define the facets in the current Examine version? e.g. in `TransformingIndexValues`?

Reply:
There's a few things you have to do for faceting to work:
1. Add a facetable field to the document for example `SortedSetDocValuesFacetField`
2. Build the document with a `FacetsConfig` - Otherwise the document will complain that it doesn't know how to index the facetable field
3. Collect the facets when searching. See the changes in `src/Examine.Lucene/Search/LuceneSearchExecutor.cs`

The `FacetField` here isn't needed for basic faceting to work as the field name has a default value that works out of the box 😄

Reply:
@nikcio okay thanks for the explanation. However I wonder it is it possible with the current Examine version without making the changes as in this PR.

Reply:
I don't think it would be very smooth because you need to hook into both the indexing process and the search. But you could try and see if you could have a `SortedSetDocValuesFacetField` indexed in a indexing event like `TransformingIndexValues`. But to search for the facets you properly have to create a custom implementation of the `LuceneSearchExecutor` where you can add the facets collection.

Reply:
I tried the `DocumentWriting` event to access `Document`:
https://github.com/Shazwazza/Examine/blob/07d0889d6d0a9ea06b9bae34a068dcb959e3322f/docs/indexing.md#luceneindexdocumentwriting

```
public void Initialize()
{
    if (!_examineManager.TryGetIndex(UmbracoIndexes.ExternalIndexName, out IIndex index))
    {
        throw new InvalidOperationException($"No index found by name {UmbracoIndexes.ExternalIndexName}");
    }

    index.TransformingIndexValues += UmbracoContextIndex_TransformingIndexValues;

    if (index is not LuceneIndex luceneIndex)
    {
        throw new InvalidOperationException($"Index {UmbracoIndexes.ExternalIndexName} is not a LuceneIndex");
    }

    luceneIndex.DocumentWriting += LuceneIndex_DocumentWriting;
}

private void LuceneIndex_DocumentWriting(object? sender, Examine.Lucene.DocumentWritingEventArgs e)
{
    if (e.ValueSet.Values.ContainsKey("duration"))
    {
        if (int.TryParse(e.ValueSet.GetValue("duration")?.ToString(), out int duration))
        {
            e.Document.Add(new SortedSetDocValuesFacetField("duration", duration.ToString()));
            e.Document.Add(new NumericDocValuesField("duration", duration));
        }
    }
}
```

However it seems it cause some errors like this:

![image](https://user-images.githubusercontent.com/2919859/207821430-e0c32a7c-e036-49d0-8e2c-2d55dd9f0c6e.png)

Reply:
Yeah that error is because the document isn't processed with a `FacetsConfig` - The reason for the change in `src/Examine.Lucene/Providers/LuceneIndex.cs`